src/SonarQube.Client/Services/SonarQubeService.cs

```diff
@@ -0,0 +1,253 @@
+            }
+
+            int currentPage = 1;
+            var allOrganizations = new List<OrganizationDTO>();
+            Result<OrganizationDTO[]> organizationsResult;
+
+            do
+            {
+                var request = new OrganizationRequest { Page = currentPage, PageSize = MaximumPageSize };
+                organizationsResult = await this.sonarqubeClient.GetOrganizationsAsync(this.connection, request, token);
+
+                if (organizationsResult.IsFailure)
+                {
+                    // TODO: something went wrong
+                }
+                else
+                {
+                    allOrganizations.AddRange(organizationsResult.Value);
+                }
+
+                currentPage++;
+            }
+            while (organizationsResult.Value.Length > 0);
+
+            return allOrganizations.Select(Organization.FromDto).ToList();
+        }
+
+        public async Task<IList<Plugin>> GetAllPluginsAsync(CancellationToken token)
+        {
+            if (!this.isConnected)
+            {
+                throw new InvalidOperationException("This operation expects the service to be connected.");
+            }
+
+            var pluginsResult = await this.sonarqubeClient.GetPluginsAsync(this.connection, token);
+            if (pluginsResult.IsFailure)
+            {
+                // TODO: something went wrong
+            }
+
+            return pluginsResult.Value.Select(Plugin.FromDto).ToList();
+        }
+
+        public async Task<IList<Project>> GetAllProjects(string organizationKey, CancellationToken token)
+        {
+            if (!this.isConnected)
+            {
+                throw new InvalidOperationException("This operation expects the service to be connected.");
+            }
+
+            if (organizationKey == null) // Orgs are not supported or not enabled
+            {
+                var projectsResult = await this.sonarqubeClient.GetProjectsAsync(this.connection, token);
+                if (projectsResult.IsFailure)
+                {
+                    // TODO: something went wrong
+                }
+
+                return projectsResult.Value.Select(Project.FromDto).ToList();
+            }
+
+            int currentPage = 1;
+            var allProjects = new List<ComponentDTO>();
+            Result<ComponentDTO[]> componentsResult;
+
+            do
+            {
+                var request = new ComponentRequest { Page = currentPage, PageSize = MaximumPageSize };
+                componentsResult = await this.sonarqubeClient.GetComponentsSearchProjectsAsync(this.connection,
+                    request, token);
+
+                if (componentsResult.IsFailure)
+                {
+                    // TODO: something went wrong
+                }
+                else
+                {
+                    allProjects.AddRange(componentsResult.Value);
+                }
+
+                currentPage++;
```

Comment:
When does the cancellation token get checked (particularly in the calls that get pages of data)?

Reply:
Cancellation token is checked at every call by the client but this is a fair comment that I could add a check too.